apps/web/components/dashboard-components/project-details/custom-llm-parameters-editor.tsx

```diff
@@ -0,0 +1,650 @@
+    }
+
+    // Get the current provider and model
+    const currentProvider =
+      selectedProvider || project.defaultLlmProviderName || "openai";
+    const currentModel = selectedModel || project.defaultLlmModelName || "";
+
+    // Create nested structure with provider -> model -> parameters
+    // Preserve existing parameters for other providers and models
+    const existingParams = project.customLlmParameters || {};
+
+    let customLlmParameters: z.infer<typeof customLlmParametersSchema> | null =
+      null;
+
+    if (Object.keys(parametersObject).length > 0) {
+      // Add/update parameters for current provider and model
+      customLlmParameters = {
+        ...existingParams,
+        [currentProvider]: {
+          ...existingParams[currentProvider],
+          [currentModel]: parametersObject,
+        },
+      };
+    } else if (existingParams[currentProvider]?.[currentModel]) {
+      // Remove parameters for current provider and model but keep others
+      const providerParams = existingParams[currentProvider];
+      const { [currentModel]: _, ...otherModels } = providerParams;
+
+      if (Object.keys(otherModels).length > 0) {
+        // Keep other models for this provider
+        customLlmParameters = {
+          ...existingParams,
+          [currentProvider]: otherModels,
+        };
+      } else {
+        // Remove entire provider if no models left
+        const { [currentProvider]: _, ...otherProviders } = existingParams;
+        customLlmParameters =
+          Object.keys(otherProviders).length > 0 ? otherProviders : null;
+      }
+    } else {
+      // No changes needed, keep existing
+      customLlmParameters =
+        Object.keys(existingParams).length > 0 ? existingParams : null;
+    }
+
+    updateProject.mutate({
+      projectId: project.id,
+      customLlmParameters,
+    });
+  };
+
+  const handleCancel = () => {
+    // Reset to original values
+    if (project?.customLlmParameters) {
+      // Use selectedProvider if provided, otherwise fall back to project's default provider
+      const currentProvider =
+        selectedProvider || project.defaultLlmProviderName;
+      // Use selectedModel if provided, otherwise fall back to project's default model
+      const currentModel = selectedModel || project.defaultLlmModelName;
+
+      if (currentProvider && currentModel) {
+        const providerParams = project.customLlmParameters[currentProvider];
+        const modelParams = providerParams?.[currentModel];
+
+        if (modelParams && typeof modelParams === "object") {
+          const entries = Object.entries(modelParams).map(([key, value]) => ({
+            key,
+            value: String(value),
+            type:
+              typeof value === "boolean"
+                ? "boolean"
+                : typeof value === "number"
+                  ? "number"
+                  : "string",
+          })) as ParameterEntry[];
+          setParameters(entries);
+        } else {
```

Comment:
I'm glad to see the `else` cases but all of this feels very repetitive. Can you get there by using more `??` and `?.`? kind of like how I simplified the save case?